### ISSUE TITLE
Excessive number of log segments generated on one partition

### ISSUE THREAD

**Issue by strawhat5** [zeebe-io/zeebe#5063]
**Describe the bug**
We have come across an issue where we are seeing an excessive number of log segments getting generated for one of the partitions (_cluster: 3 broker, 3 partitions, standalone gateway, **no exporters**_). The number of segments generated for one of the partition has reached around 300, with 512MB (logSegmentSize) each. That is almost 150GB of data in our PVC (per broker), and we have started seeing some other issues as well.

We are seeing continuous errors `Failed to obtain a pending snapshot directory for position x` for the last 1 day. And we are also seeing in Grafana that the Snapshot Operation has stopped since yesterday.
<img width="562" alt="Screenshot 2020-07-27 at 10 41 24 PM" src="https://user-images.githubusercontent.com/10865861/88570990-9bef3280-d05a-11ea-9b09-bea40bc36df4.png">
<img width="561" alt="Screenshot 2020-07-27 at 10 41 39 PM" src="https://user-images.githubusercontent.com/10865861/88570999-9f82b980-d05a-11ea-96b6-d2d93b050524.png">

We have also seen one other error coming in the broker logs:
```
2020-07-26 17:44:44.667 [Broker-1-StreamProcessor-3] [Broker-1-zb-actors-1] [31mERROR[m io.zeebe.broker.workflow - Expected to find job with key 6755399445248597, but no job found
```

**To Reproduce**
On investigating the log segments, we found these errors in most of the segments:
```
Expected to time out activated job with key '6755399445248597', but it must be activated firstŽ¨deadlineÏs‹éH‡¦worker±logistics§retries¤type±logisticscustomHeaders€©variablesÄ€¬errorMessageÚÅjavax.persistence.PersistenceException: org.hibernate.exception.JDBCConnectionException: Unable to acquire JDBC Connection
```
After seeing these errors, we cancelled their corresponding workflow instances, expecting the compaction to get triggered. But the compaction did not get triggered and we still have the same number of `raft-partition*.log` files (around 300) present for that partition. At least after cancelling these instances, the data growth has reduced significantly, but still growing gradually (and without creating new segments! Talk about all the weird issues coming together! 😅 ).

**Expected behavior**
As per the [doc](https://docs.zeebe.io/operations/resource-planning.html#event-log), `Log segment deletion will begin as soon as the max number of snapshots has been reached`. But the segments have not been deleted and the data is continuously growing.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
2020-07-27 15:03:20.979 [Broker-0-SnapshotDirector-2] [Broker-0-zb-fs-workers-1] [33mWARN [m io.zeebe.logstreams.snapshot - Failed to obtain a pending snapshot directory for position 11050954632720
2020-07-27 15:11:51.468 [] [raft-server-0-raft-partition-partition-1] [33mWARN [m io.atomix.raft.roles.LeaderAppender - RaftServer{raft-partition-partition-1} - AppendRequest{term=182, leader=0, prevLogIndex=54210760, prevLogTerm=182, entries=0, commitIndex=54722997} to 2 failed: java.util.concurrent.CompletionException: java.util.concurrent.TimeoutException: Request type raft-partition-partition-1-append timed out in 5000 milliseconds
java.util.concurrent.CompletionException: java.util.concurrent.TimeoutException: Request type raft-partition-partition-1-append timed out in 5000 milliseconds
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$executeOnPooledConnection$17(NettyMessagingService.java:482) ~[atomix-cluster-0.23.4.jar:0.23.4]
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30) ~[guava-28.2-jre.jar:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$executeOnPooledConnection$18(NettyMessagingService.java:482) ~[atomix-cluster-0.23.4.jar:0.23.4]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
	at io.atomix.cluster.messaging.impl.AbstractClientConnection$Callback.timeout(AbstractClientConnection.java:163) ~[atomix-cluster-0.23.4.jar:0.23.4]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: java.util.concurrent.TimeoutException: Request type raft-partition-partition-1-append timed out in 5000 milliseconds
	... 7 more
2020-07-27 15:13:27.022 [Broker-0-SnapshotDirector-1] [Broker-0-zb-fs-workers-0] [33mWARN [m io.zeebe.logstreams.snapshot - Failed to obtain a pending snapshot directory for position 11987254449144
2020-07-27 15:18:21.011 [Broker-0-SnapshotDirector-2] [Broker-0-zb-fs-workers-1] [33mWARN [m io.zeebe.logstreams.snapshot - Failed to obtain a pending snapshot directory for position 11050956241496
```

</p>
</details>

**Environment:**
 - Zeebe Version: `0.23.4`
<details><summary>Full Stacktrace</summary>
 <p>

```
Starting broker 0 with configuration {
  "network" : {
    "host" : "0.0.0.0",
    "portOffset" : 0,
    "maxMessageSize" : "4MB",
    "advertisedHost" : "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local",
    "commandApi" : {
      "host" : "0.0.0.0",
      "port" : 26501,
      "advertisedHost" : "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local",
      "advertisedPort" : 26501,
      "address" : "0.0.0.0:26501",
      "advertisedAddress" : "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26501"
    },
    "internalApi" : {
      "host" : "0.0.0.0",
      "port" : 26502,
      "advertisedHost" : "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local",
      "advertisedPort" : 26502,
      "address" : "0.0.0.0:26502",
      "advertisedAddress" : "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26502"
    },
    "monitoringApi" : {
      "host" : "0.0.0.0",
      "port" : 9600,
      "advertisedHost" : "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local",
      "advertisedPort" : 9600,
      "address" : "0.0.0.0:9600",
      "advertisedAddress" : "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local:9600"
    },
    "maxMessageSizeInBytes" : 4194304
  },
  "cluster" : {
    "initialContactPoints" : [ "prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26502", "prod-zeebe-zeebe-broker-1.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26502", "prod-zeebe-zeebe-broker-2.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26502" ],
    "partitionIds" : [ 1, 2, 3 ],
    "nodeId" : 0,
    "partitionsCount" : 3,
    "replicationFactor" : 3,
    "clusterSize" : 3,
    "clusterName" : "zeebe-cluster",
    "gossipFailureTimeout" : 10000,
    "gossipInterval" : 250,
    "gossipProbeInterval" : 1000
  },
  "threads" : {
    "cpuThreadCount" : 2,
    "ioThreadCount" : 2
  },
  "data" : {
    "directories" : [ "/usr/local/zeebe/data" ],
    "logSegmentSize" : "512MB",
    "snapshotPeriod" : "PT15M",
    "logIndexDensity" : 100,
    "logSegmentSizeInBytes" : 536870912,
    "atomixStorageLevel" : "DISK"
  },
  "exporters" : { },
  "gateway" : {
    "network" : {
      "host" : "0.0.0.0",
      "port" : 26500,
      "minKeepAliveInterval" : "PT30S"
    },
    "cluster" : {
      "contactPoint" : "0.0.0.0:26502",
      "requestTimeout" : "PT15S",
      "clusterName" : "zeebe-cluster",
      "memberId" : "gateway",
      "host" : "0.0.0.0",
      "port" : 26502
    },
    "threads" : {
      "managementThreads" : 1
    },
    "monitoring" : {
      "enabled" : false,
      "host" : "0.0.0.0",
      "port" : 9600
    },
    "security" : {
      "enabled" : false,
      "certificateChainPath" : null,
      "privateKeyPath" : null
    },
    "enable" : false
  },
  "backpressure" : {
    "enabled" : true,
    "algorithm" : "VEGAS"
  },
  "stepTimeout" : "PT5M"
}

```
 </p>
</details>

**Comment by strawhat5** [comment]
Hey any updates here? This is happening on our prod cluster.

Any help here is appreciated.

**Comment by strawhat5** [comment]
Adding few more logs that we have observed in the broker logs:
```
2020-07-29 13:44:42.976 [] [raft-server-2-system-partition-1] WARN  io.atomix.raft.roles.LeaderAppender - RaftServer{system-partition-1} - ConfigureRequest{term=100, leader=2, index=0, timestamp=1586954715914, members=[DefaultRaftMember{id=2, type=ACTIVE, updated=2020-04-15T12:45:15.914585Z}, DefaultRaftMember{id=0, type=ACTIVE, updated=2020-04-15T12:45:15.914585Z}, DefaultRaftMember{id=1, type=ACTIVE, updated=2020-04-15T12:45:15.914585Z}]} to 0 failed: java.util.concurrent.CompletionException: java.util.concurrent.TimeoutException: Request type system-partition-1-configure timed out in 5000 milliseconds
java.util.concurrent.CompletionException: java.util.concurrent.TimeoutException: Request type system-partition-1-configure timed out in 5000 milliseconds
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$executeOnPooledConnection$17(NettyMessagingService.java:482) ~[atomix-cluster-0.23.4.jar:0.23.4]
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30) ~[guava-28.2-jre.jar:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$executeOnPooledConnection$18(NettyMessagingService.java:482) ~[atomix-cluster-0.23.4.jar:0.23.4]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
	at io.atomix.cluster.messaging.impl.AbstractClientConnection$Callback.timeout(AbstractClientConnection.java:163) ~[atomix-cluster-0.23.4.jar:0.23.4]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: java.util.concurrent.TimeoutException: Request type system-partition-1-configure timed out in 5000 milliseconds
	... 7 more
```

**zbctl status --insecure on standalone gateway:**
```
Cluster size: 3
Partitions count: 3
Replication factor: 3
Gateway version: 0.23.4
Brokers:
  Broker 0 - prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26501
    Version: 0.23.4
    Partition 1 : Leader
    Partition 2 : Follower
    Partition 3 : Follower
  Broker 1 - prod-zeebe-zeebe-broker-1.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26501
    Version: 0.23.4
    Partition 1 : Follower
    Partition 2 : Leader
    Partition 3 : Leader
  Broker 2 - prod-zeebe-zeebe-broker-2.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26501
    Version: 0.23.4
    Partition 2 : Follower
    Partition 3 : Follower
```

We can see that the status of Partition 1 in Broker 2 is missing, could that be the reason for these errors `[raft-server-2-system-partition-1] WARN  io.atomix.raft.roles.LeaderAppender`? And since one of the follower is missing, it is not able to trigger a snapshot?

**Comment by korthout** [comment]
Hi @strawhat5,

Sorry for taking so long to reply. I don't immediately have an answer for you, but I'm looking into it right now.

**Comment by korthout** [comment]
At this time, we are unclear on what causes this. However, we feel it might be benefitial to restart broker 2.

**Comment by korthout** [comment]
At this time, we are unclear on what causes this. However, we feel it might be beneficial to restart broker 2. Please note that, since broker 2 has not been following partition 1 for a long time, it will also take a long time to replicate all events to that partition from the last snapshot. Restart might take longer than usual.

**Comment by strawhat5** [comment]
Yeah that is one of the issue that we want to avoid. As mentioned in my issue:

> The number of segments generated for one of the partition has reached around 300, with 512MB (logSegmentSize) each. That is almost 150GB of data in our PVC (per broker)

So already it is taking around an hour to restart a broker. I know restarting can probably help in assigning all partitions correctly to broker 2, but the bigger problem is how to get rid of the all these raft-partition log segments. Is it safe to delete the older log segments (I read somewhere that the raft-partition logs are monotonically increasing)? Or would that impact existing running instances as well?

If it is safe to delete those logs, then we can restart after deleting those log segments. It would be much more quicker for us.

**Comment by korthout** [comment]
No, I don't believe it is safe to delete those logs. 

I'm sorry, I'm unable to provide better solutions. We've been trying to think of additional information that might help us to get to the bottom of the issue, but at this time we can't think of anything that will help us out here. It seems that somehow 2 brokers are unable to communicate about a specific partition, but what the reason is is unclear. We've been looking at the code dealing with this. So far we've thought of the following:

- This may happen when the database is closed
- This may happen when logs have not been exported, but that is only the case when there are exporters and there are none in your config
- This may happen when the broker is not making any progress, but as far as I see in your provided logs, the index increased between the 2 snapshots that failed to be created
- This may happen when there is a problem with committing raft events, and yes we see a warning/error log about this, but as long as at least 1 follower is committing, this should be fine as well.

To summarise, it's difficult to pinpoint the problem, and a fairly easy and 'relatively' quick solution is restarting the unhealthy broker.

**Comment by strawhat5** [comment]
Ah I see your point. I will try to restart it tonight, I'm sure restart is going to take a long time. 😔 

Maybe we need better metrics regarding RocksDB (currently we don't have any way to know the internal state of DB), and some way to forcefully trigger a compaction. Something like: `zbctl compact`

**Comment by korthout** [comment]
Do you have the RocksDB tab in the Zeebe grafana dashboard? It offers quite some info, although not everything. For example, you can see how often the RocksDB data is compacted there. Note that RocksDB compaction is something else than zeebe's log compaction. RocksDB is only used to store the workflow state. When we can delete the data for a workflow (e.g. when it was completed), then we can delete it from the RocksDB storage. When RocksDB's compaction runs it will then cleanup that data. This is something else than compacting the Zeebe log. Events in the zeebe log can be deleted (i.e. the log can be compacted) when the event is part of a snapshot (and the event has been exported by all exporters). So to summarise: RocksDB stores workflow data and compaction reduces its size by cleaning up, Zeebe's log contains events and compaction of the log means deleting events that are no longer needed for Zeebe's operation.

Finally, on a command to trigger compaction, I doubt this adds anything useful. In your case, we are unable to compact the log because no new snapshots are taken (although, we do try to take them every 15 minutes). Only events that have been processed and for which their changes are part of a snapshot, can actually be deleted. So in your situation, if we would have an implemented `zbctl compact`, it would result in no changes to the log. In your case, we need to fix that snapshots are taken again. Only then can the log be compacted.

**Comment by strawhat5** [comment]
Thanks for this detailed explanation @korthout ! It alleviates a lot of doubts regarding log compaction and snapshot compaction. :)

Just a couple of questions regarding your explanation:

> compaction of the log means deleting events that are no longer needed for Zeebe's operation.
> Only events that have been processed and for which their changes are part of a snapshot, can actually be deleted.

I want to cover these 2 points using an example:
<img width="502" alt="Screenshot 2020-07-31 at 3 59 23 PM" src="https://user-images.githubusercontent.com/10865861/89026788-ecaaa800-d346-11ea-9b88-b5f965683630.png">

1. Suppose that an instance has been created for the above workflow.
2. Task1 is completed.
3. After step 2, lots of other workflows instances get created (for an unrelated workflow), and they get completed as well. Task2 is still not completed.
4. My `raft-partition-0-1.log` gets rolled over to `raft-partition-0-2.log` due to events from all other WF instances. What I mean is that now there are more than 1 log-segment for a partition (as log compaction has not taken place yet).
5. Now snapshot operation takes place, followed by RocksDB compaction. This deletes the data of all completed WF instances from RocksDB. (Now my RocksDB size is also reduced).

Now after Step 5 is finished, will the log compaction be triggered?
> compaction of the log means deleting events that are no longer needed for Zeebe's operation.

Here what do you mean by **_events_**?
- Does event mean the state of a workflow instance? -> In that case `raft-partition-0-1.log` will not be deleted until instance is either completed or cancelled.
- Does it mean regular events like JobCompletionRequest? -> In this case `raft-partition-0-1.log` should be deleted.

> Only events that have been processed and for which their changes are part of a snapshot, can actually be deleted.

Here what do you mean by **_processed_**? -> This is actually related to your definition of events. If events are indeed state of a workflow instance, then I guess _processed_ would mean completion/cancellation of the instance.

**Comment by korthout** [comment]
Ah, I'm not sure if I can answer everything in your comment, but I think it makes sense to explain what I mean by events.

> Internally, Zeebe is implemented as a collection of stream processors working on record streams (partitions)

https://docs.zeebe.io/basics/internal-processing.html

When I talk about events, then those are a type of record. Commands would be another. So basically, Zeebe has a log stream (or perhaps better: a log stream of records) and Zeebe consists of a collection of processors. Each processor changes the internal state of a workflow. That internal state is stored in RocksDB (the log stream is not). An event is **processed** when a stream processor has processed the event, which probably changed the internal state and possibly created new 'follow-up' events. 

Consider a workflow instance starting at a start event. For the start event, we will produce an `ELEMENT_ACTIVATING` event. The stream processor `StartEventProcessor` will handle this event. It changes the internal state of the start event element to `ELEMENT_ACTIVATED` and writes a 'follow-up' event (`ELEMENT_ACTIVATED` event) on the log stream. Now that event can be processed, again changing the internal state and potentially producing new follow-up events. (you can read more about the lifecycle of workflow elements here: https://docs.zeebe.io/reference/workflow-lifecycles.html)

Sometimes we can delete things from the state (this generally happens when a workflow completes) and RocksDB can then properly clean it up in one of its compactions. Next to that, we can clean up our own log stream, since already processed events already changed the state. If we also backed up that state in the form of a snapshot AND replicated that snapshot to a follower AND we've made sure the event is exported to all exporters (for example for traceability) then Zeebe no longer needs the event in the stream and we can delete it. This step we do in our own compaction.

I hope this helps, if it's still unclear. Please let me know.

**Comment by strawhat5** [comment]
Right, that's a great explanation! Thanks @korthout for taking the effort to jot it down together. 🙂 

Now, coming back to the problem in hand. I did restart the broker node 2, but not to much avail. It did stop the LeaderAppender Raft warnings: `Caused by: java.util.concurrent.TimeoutException: Request type system-partition-1-configure timed out in 5000 milliseconds`. And on doing `zbctl status` I am now able to see the status of all 3 brokers correctly.

But I am still seeing these errors in my broker logs, and the snapshot operation is not getting triggered:
```
2020-07-31 03:59:27.829 [Broker-1-SnapshotDirector-2] [Broker-1-zb-fs-workers-0] WARN  io.zeebe.logstreams.snapshot - Failed to obtain a pending snapshot directory for position 11239935936752
2020-07-31 04:14:27.846 [Broker-1-SnapshotDirector-2] [Broker-1-zb-fs-workers-0] WARN  io.zeebe.logstreams.snapshot - Failed to obtain a pending snapshot directory for position 11239936370000
2020-07-31 04:29:27.868 [Broker-1-SnapshotDirector-2] [Broker-1-zb-fs-workers-0] WARN  io.zeebe.logstreams.snapshot - Failed to obtain a pending snapshot directory for position 11239936749872
2020-07-31 04:44:27.886 [Broker-1-SnapshotDirector-2] [Broker-1-zb-fs-workers-1] WARN  io.zeebe.logstreams.snapshot - Failed to obtain a pending snapshot directory for position 11239937150088
2020-07-31 04:59:27.909 [Broker-1-SnapshotDirector-2] [Broker-1-zb-fs-workers-0] WARN  io.zeebe.logstreams.snapshot - Failed to obtain a pending snapshot directory for position 11239937656936
```

As you can see, these errors are coming exactly every 15 minutes (which is the default snapshot interval). So Zeebe is probably trying to trigger a snapshot operation but failing due to these errors.

**Comment by korthout** [comment]
@strawhat5 happy to help in what ever way I can. 🙂 

Sorry to hear this did not resolve the issue (fully), but its very interesting that it didn't while solving the raft timeout. 

About the remaining warnings: You are right that Zeebe tries to make a new snapshot every x minutes (default: 15 min) and that is what we see failing here. Question is still: why does it fail?

Can you perhaps make some screenshots of the grafana Zeebe dashboard? Maybe something stands out to me in there.

**Comment by strawhat5** [comment]
Attaching screenshots from _Snapshots_ and _Atomix_ section below. Let me know if you need screenshots from other sections as well, or any other related data.
<img width="1186" alt="Screenshot 2020-07-31 at 7 31 51 PM" src="https://user-images.githubusercontent.com/10865861/89042730-05c25180-d365-11ea-8819-f7d9756dbac5.png">
<img width="1181" alt="Screenshot 2020-07-31 at 7 32 06 PM" src="https://user-images.githubusercontent.com/10865861/89042746-0a870580-d365-11ea-9619-d43ebeb643cc.png">
<img width="1181" alt="Screenshot 2020-07-31 at 7 32 30 PM" src="https://user-images.githubusercontent.com/10865861/89042748-0b1f9c00-d365-11ea-8cf5-2fefeddf5c37.png">
<img width="1186" alt="Screenshot 2020-07-31 at 7 32 45 PM" src="https://user-images.githubusercontent.com/10865861/89042751-0c50c900-d365-11ea-9308-923d00444a05.png">
<img width="1184" alt="Screenshot 2020-07-31 at 7 34 44 PM" src="https://user-images.githubusercontent.com/10865861/89042754-0e1a8c80-d365-11ea-8b97-ebc2a65bbe6a.png">

**Comment by korthout** [comment]
Could you perhaps run the following on each broker and print the output in a comment here?

```
ls -R -l data
```

We don't know for sure, but we have a feeling it might be useful to see a bit of meta data on the data stored for each partition.

**Comment by korthout** [comment]
Also could you screenshot the partition health metrics dashboard?

**Comment by strawhat5** [comment]
Attached `ls -R -l data` for each broker below.

[zb0.txt](https://github.com/zeebe-io/zeebe/files/5007607/zb0.txt)
[zb1.txt](https://github.com/zeebe-io/zeebe/files/5007609/zb1.txt)
[zb2.txt](https://github.com/zeebe-io/zeebe/files/5007610/zb2.txt)

Attached partition health metrics below.
<img width="1204" alt="Screenshot 2020-07-31 at 8 32 46 PM" src="https://user-images.githubusercontent.com/10865861/89048333-53db5300-d36d-11ea-8ab6-eb5240dabcdb.png">

**Comment by korthout** [comment]
Thanks for all this info. I don't immediately see anything big. Looks like none of the partitions have had snapshots taken since 26-07-2020 06:20. This is different from what I originally thought, namely that only 1 partition wasn't taking snapshots. I need to spend some more time on this, but I hope you understand that weekend is upon us so I need to postpone this for now. Let's see after the weekend whether we can solve this.

**Comment by strawhat5** [comment]
Hey sure @korthout ! Thanks for all your help here. 🙂 

I don't think our cluster is unstable right now (other than this issue), so we should be able to get by this weekend safely. Looking forward to continue on this next week. :)

**Comment by npepinpe** [comment]
Prioritized as critical for root cause analysis - once we've figured out what the cause is let's review the priority.

**Comment by korthout** [comment]
@strawhat5 we're bringing the big guns in with @Zelldon 🙂

**Comment by Zelldon** [comment]
Hey @strawhat5 

in order to help you here I need to ask some more questions.

I will try to categorize them a bit.

**Investigation:**

 * Do you have still the logs accessible? 
   * I would like to see the logs from Broker 0, 1 and 2 around the 16.07.2020 ~ 8 pm to 0 am. It seems for me that around this time the last compaction happens, since the journal is increasing since then. Broker 0 has as first segment `166.log` with timestamp 16.07.2020 20:58, Broker 1 has as first segment `167.log` with timestamp 16.07.2020 20:58 and Broker-2 has as first segment `162.log` with timestamp 16.07.2020 20:29.

E.g from Broker-0:
```
/usr/local/zeebe/data/raft-partition/partitions/3:
total 182178776
drwxr-xr-x 2 root root      4096 May 20 06:34 pending
drwxr-xr-x 3 root root      4096 Jul 26 04:41 pushed-pending
-rw-r--r-- 1 root root 536870912 Jul 16 20:58 raft-partition-partition-3-166.log
-rw-r--r-- 1 root root 536870912 Jul 16 22:12 raft-partition-partition-3-167.log
```

 * Do you have seen anything unexpected, except the big segment count? Do you experience any throughput/latency issues? 
 * Do you think you can compress one of the data folders and make it available for us? This would make it much easier to investigate this issue.


**Reproducing:**

 * Do you have an idea to reproduce this? 
 * What kind of load do you running? 
 * How do you deploy zeebe?
 * How does your workflows look like?

**Comment by strawhat5** [comment]
Hey Chris, I will try to answer your questions in order.

> Do you have still the logs accessible?

Unfortunately, no. First time we did notice it was around July 25th, when one of the broker restarted and started going into OOM error. We had a 4core/16GB machine earlier. We upgraded it to 8core/32GB machine to stop the OOM error, and let the broker successfully restart. That was the first time when we saw huge number of partition-log files accumulated in our PVC. We don't have any logs earlier than July 25th.

> Do you have seen anything unexpected, except the big segment count? Do you experience any throughput/latency issues?

We have not seen any latency/throughput issues. The only other issue we saw was in elasticsearch exporter, where the exporter was trailing behind stream processor by around 3 days but gradually moving ahead (we could see in kibana that the records were getting exported but we could only see the indexed records for 3 days earlier - and we verified that there were no issues on the side of elasticsearch cluster). We removed the exporter (from `broker.yaml`) during our July 25th update.

> Do you think you can compress one of the data folders and make it available for us? This would make it much easier to investigate this issue.

The data folder (for each broker) has reached over 150GB in size. So it would be difficult to get it completely. But I can definitely provide you some parts of it, if you could tell me which specific folders you would be interested in. Maybe I provide you one of the `raft-partition*.log` file along with other folders. We investigated some of the `raft-partition*.log` files, and they are mostly filled with this error (we identified the cause of this error and cancelled error-prone workflow instances):
```
Expected to time out activated job with key '6755399445248597', but it must be activated firstŽ¨deadlineÏs‹éH‡¦worker±logistics§retries¤type±logisticscustomHeaders€©variablesÄ€¬errorMessageÚÅjavax.persistence.PersistenceException: org.hibernate.exception.JDBCConnectionException: Unable to acquire JDBC Connection
```

> Did you do a version update in between?

No we did not. We have been using `0.23.4` till date.

**Comment by strawhat5** [comment]
**Reproducing**:
I don't have any exact way to reproduce it, but on root causing the issue I could outline a potential way for it.

> Do you have an idea to reproduce this?

We have a simple workflow with a service task (attached below) fetching a row from DB. Some attributes of that task worker:
`timeout`: 20 seconds ([Job Timeout](https://docs.zeebe.io/reference/grpc.html?highlight=timeout#activatejobs-rpc))
`requestTimeout`: 20 seconds
`jdbcConnectionTimeout`: 30 seconds
<img width="496" alt="Screenshot 2020-08-03 at 6 51 23 PM" src="https://user-images.githubusercontent.com/10865861/89187078-5bdc0280-d5ba-11ea-9230-3975a6cc2d69.png">
The worker was activating the task and getting a JDBC connection timeout error every single time. And by the time it could throw `org.hibernate.exception.JDBCConnectionException`, the job was getting timed out, and being assigned to another worker thread. This continued for 15 days straight (from what we can see in Kibana). Attaching a screenshot below.

<img width="1096" alt="Screenshot 2020-08-03 at 6 47 04 PM" src="https://user-images.githubusercontent.com/10865861/89186754-dce6ca00-d5b9-11ea-8485-be2d58528bd1.png">

> What kind of load do you running?

There is not much load. We were running approximately 1000 instances per day, of below workflow. And some of those instances were facing this issue.

> How do you deploy zeebe?

In Kubernetes using our own Helm charts.

> How does your workflows look like?

Attached above. We have one other workflow as well, but that has been running smoothly, has not caused any exceptions yet. That is a simple timer followed by a service task.

**Comment by korthout** [comment]
Thanks for all this info. We immediately have some follow-up questions:

- These JDBCConnectionExceptions seem to stop around the same time that compaction seems to stop from our perspective. Do you remember whether you did anything around that time that might be interesting to us?
- Do you remember when you upgraded to `0.23.4`?

**Comment by strawhat5** [comment]
Hey @korthout that's a good observation! I have verified that we did not change anything on either Zeebe cluster or on the worker thread. But your observation holds true as I just verified it from our simple monitor, and the last update time matches!
<img width="1253" alt="Screenshot 2020-08-03 at 7 35 05 PM" src="https://user-images.githubusercontent.com/10865861/89191498-a3fe2380-d5c0-11ea-96ad-007f568191e3.png">


Alas we do not have the broker logs of that time, which could have helped better.
We upgraded our cluster to `0.23.4` on July 10th.

**Comment by Zelldon** [comment]
Hey @strawhat5 

this is interesting and thanks for the details! So you actually used also simple monitor, which means you have enabled the hazelcast exporter as well right? Could you please tell us exactly how do you disabled the exporters? Might be the issue that it is not correctly removed from the internal state.

**Comment by Zelldon** [comment]
Ok I think I found the issue. I need to verify this via a test. I will do it tomorrow.

If you take a look at this line https://github.com/zeebe-io/zeebe/blob/develop/broker/src/main/java/io/zeebe/broker/system/partitions/ZeebePartition.java#L548 you can see that we not installing the `ExporterDirector`, when there is no exporter configured. This is only problematic, if you have configured exporters before. This means we still have the exporters referenced in our state, which means we will not clean up our log and wait for the referenced exporters until they exported all the records (which will never happen of course).

To work around this just enable an exporter again, then you should be able to compact again. 

I will try to reproduce it tomorrow and create a separate bug issue for it. Thanks for reporting and your help.

**Comment by npepinpe** [comment]
Great catch! I'm hoping to have a patch release next week so we can get that in there :)

**Comment by strawhat5** [comment]
Great! 😃

I vaguely remember having a discussion on this with @npepinpe. :) I thought just removing the exporters from `broker.yaml` and restarting the broker should trigger a compaction of logs. I think I assumed that state is decoupled from exporter position.

Just to confirm, if I configure the exporter in `broker.yaml` and restart the broker again, that should do thr trick right? Also, do I need to configure both hazelcast and elastic exporters (I had both earlier), or just a single exporter would be enough?

**Comment by npepinpe** [comment]
That is the expected behavior, so if Zell is right this is definitely a bug 👍

**Comment by Zelldon** [comment]
Issue https://github.com/zeebe-io/zeebe/issues/5105 and PR https://github.com/zeebe-io/zeebe/pull/5106 are open. After merging the PR, backporting and releasing I will close this.

**Comment by strawhat5** [comment]
Thanks Chris! :)

Can you please help me answer these questions? (for `0.23.4` version)
>Just to confirm, if I configure the exporter in `broker.yaml` and restart the broker again, that should do the trick right?
Also, do I need to configure both hazelcast and elastic exporters (I had both configured earlier), or just a single exporter would be enough (does the state maintain a single position for all exporters, or individual exporter positions?) ?

**Comment by Zelldon** [comment]
Sorry I have forgot to answer this before.

Yes, one exporter should be enough. It should work then. The broker maintains positions per exporter. If some exporter is removed it should remove the related exporter and his last position from the state and keep the others.

The bug was that if no exporter is configured the clean up was not happening.

**Comment by strawhat5** [comment]
Got it, thanks! I will try it out in our cluster.

**Comment by korthout** [comment]
Great work @Zelldon!

**Comment by strawhat5** [comment]
Hey, **just an update**: we reconfigured one exporter and restarted one of the brokers (which was a partition leader), and we can see snapshot operations have started again. Atomix log compaction is also happening on that partition. 🎉  We will restart the other brokers tonight and will post an update here.

**And a follow-up question**: we saw these errors in one of the brokers:
```
java.lang.IllegalStateException: LeaderRole is closed and cannot be used as appender
Failed to install follower partition 2 java.lang.RuntimeException: Actor is closed
```
<details><summary>Full Stacktrace</summary>
 <p>

```
2020-08-04 23:39:10.844 [Broker-2-SnapshotDirector-2] [Broker-2-zb-fs-workers-0] DEBUG io.zeebe.logstreams.snapshot - Replicate snapshot chunk /usr/local/zeebe/data/raft-partition/partitions/2/snapshots/50514399-209-1596584349885/028389.sst
2020-08-04 23:41:08.498 [Broker-2-ZeebePartition-2] [Broker-2-zb-actors-2] DEBUG io.zeebe.broker.system - Removing leader partition services for partition PartitionId{id=2, group=raft-partition}
2020-08-04 23:41:08.498 [Broker-2-ZeebePartition-2] [Broker-2-zb-actors-2] DEBUG io.zeebe.broker.system - Closing Broker-2-Exporter-2
2020-08-04 23:41:08.498 [Broker-2-ZeebePartition-2] [Broker-2-zb-actors-2] DEBUG io.zeebe.broker.system - Partition role transitioning from LEADER to FOLLOWER
2020-08-04 23:41:08.498 [] [raft-server-2-raft-partition-partition-2] ERROR io.zeebe.logstreams - Failed to append block with last event position 11506225119720.
java.lang.IllegalStateException: LeaderRole is closed and cannot be used as appender
	at io.atomix.raft.roles.LeaderRole.safeAppendEntry(LeaderRole.java:1422) ~[atomix-cluster-0.23.4.jar:0.23.4]
	at io.atomix.raft.roles.LeaderRole.lambda$appendEntry$32(LeaderRole.java:1407) ~[atomix-cluster-0.23.4.jar:0.23.4]
	at io.atomix.utils.concurrent.SingleThreadContext$WrappedRunnable.run(SingleThreadContext.java:188) [atomix-utils-0.23.4.jar:0.23.4]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
2020-08-04 23:41:08.498 [] [raft-server-2-raft-partition-partition-2] ERROR io.zeebe.logstreams - Failed to append block with last event position 11506225119920.
java.lang.IllegalStateException: LeaderRole is closed and cannot be used as appender
	at io.atomix.raft.roles.LeaderRole.safeAppendEntry(LeaderRole.java:1422) ~[atomix-cluster-0.23.4.jar:0.23.4]
	at io.atomix.raft.roles.LeaderRole.lambda$appendEntry$32(LeaderRole.java:1407) ~[atomix-cluster-0.23.4.jar:0.23.4]
	at io.atomix.utils.concurrent.SingleThreadContext$WrappedRunnable.run(SingleThreadContext.java:188) [atomix-utils-0.23.4.jar:0.23.4]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
2020-08-04 23:41:08.499 [Broker-2-LogAppender-2] [Broker-2-zb-actors-0] ERROR io.zeebe.logstreams - Actor Broker-2-LogAppender-2 failed in phase STARTED.
java.lang.IllegalStateException: LeaderRole is closed and cannot be used as appender
	at io.atomix.raft.roles.LeaderRole.safeAppendEntry(LeaderRole.java:1422) ~[atomix-cluster-0.23.4.jar:0.23.4]
	at io.atomix.raft.roles.LeaderRole.lambda$appendEntry$32(LeaderRole.java:1407) ~[atomix-cluster-0.23.4.jar:0.23.4]
	at io.atomix.utils.concurrent.SingleThreadContext$WrappedRunnable.run(SingleThreadContext.java:188) ~[atomix-utils-0.23.4.jar:0.23.4]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
	at java.lang.Thread.run(Unknown Source) ~[?:?]
2020-08-04 23:41:08.500 [Broker-2-LogStream-2] [Broker-2-zb-actors-0] INFO  io.zeebe.logstreams - Close appender for log stream logstream-raft-partition-partition-2
2020-08-04 23:41:08.500 [Broker-2-ZeebePartition-2] [Broker-2-zb-actors-0] ERROR io.zeebe.broker.system - logstream-raft-partition-partition-2 failed, marking it as unhealthy
2020-08-04 23:41:08.500 [logstream-raft-partition-partition-2-write-buffer] [Broker-2-zb-actors-2] DEBUG io.zeebe.dispatcher - Dispatcher closed
2020-08-04 23:41:08.500 [Broker-2-LogStream-2] [Broker-2-zb-actors-2] INFO  io.zeebe.logstreams - On closing logstream logstream-raft-partition-partition-2 close 3 readers
2020-08-04 23:41:08.501 [Broker-2-LogStream-2] [Broker-2-zb-actors-2] INFO  io.zeebe.logstreams - Close log storage with name logstream-raft-partition-partition-2
2020-08-04 23:41:08.501 [Broker-2-ZeebePartition-2] [Broker-2-zb-actors-0] DEBUG io.zeebe.broker.system - Detected unhealthy components. The current health status of components: {Raft-2=HEALTHY, Broker-2-StreamProcessor-2=HEALTHY, logstream-raft-partition-partition-2=UNHEALTHY}
2020-08-04 23:41:08.501 [Broker-2-HealthCheckService] [Broker-2-zb-actors-0] ERROR io.zeebe.broker.system - Partition-2 failed, marking it as unhealthy
2020-08-04 23:41:09.607 [Broker-2-Exporter-2] [Broker-2-zb-fs-workers-1] DEBUG io.zeebe.broker.exporter - Closed exporter director 'Broker-2-Exporter-2'.
2020-08-04 23:41:09.607 [Broker-2-ZeebePartition-2] [Broker-2-zb-actors-0] DEBUG io.zeebe.broker.system - Closed Broker-2-Exporter-2 successfully
2020-08-04 23:41:09.607 [Broker-2-ZeebePartition-2] [Broker-2-zb-actors-0] DEBUG io.zeebe.broker.system - Closing Broker-2-SnapshotDirector-2
2020-08-04 23:41:09.608 [Broker-2-SnapshotDirector-2] [Broker-2-zb-fs-workers-1] ERROR io.zeebe.logstreams.snapshot - Unexpected error on retrieving commit position
2020-08-04 23:41:09.608 [Broker-2-ZeebePartition-2] [Broker-2-zb-actors-2] DEBUG io.zeebe.broker.system - Unexpected exception on closing Broker-2-SnapshotDirector-2
java.lang.RuntimeException: Actor is closed
	at io.zeebe.util.sched.ActorJob.failFuture(ActorJob.java:166) ~[zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ClosedQueue.offer(ClosedQueue.java:84) ~[zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ClosedQueue.offer(ClosedQueue.java:15) ~[zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorTask.submit(ActorTask.java:95) ~[zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorControl.call(ActorControl.java:135) ~[zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.logstreams.impl.log.LogStreamImpl.getCommitPositionAsync(LogStreamImpl.java:105) ~[zeebe-logstreams-0.23.4.jar:0.23.4]
	at io.zeebe.engine.processor.AsyncSnapshotDirector.lambda$closeAsync$1(AsyncSnapshotDirector.java:108) ~[zeebe-workflow-engine-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:28) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:76) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.23.4.jar:0.23.4]
2020-08-04 23:41:09.609 [Broker-2-ZeebePartition-2] [Broker-2-zb-actors-2] ERROR io.zeebe.broker.system - Unexpected exception on removing leader partition!
java.lang.RuntimeException: Actor is closed
	at io.zeebe.util.sched.ActorJob.failFuture(ActorJob.java:166) ~[zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ClosedQueue.offer(ClosedQueue.java:84) ~[zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ClosedQueue.offer(ClosedQueue.java:15) ~[zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorTask.submit(ActorTask.java:95) ~[zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorControl.call(ActorControl.java:135) ~[zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.logstreams.impl.log.LogStreamImpl.getCommitPositionAsync(LogStreamImpl.java:105) ~[zeebe-logstreams-0.23.4.jar:0.23.4]
	at io.zeebe.engine.processor.AsyncSnapshotDirector.lambda$closeAsync$1(AsyncSnapshotDirector.java:108) ~[zeebe-workflow-engine-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:28) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:76) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.23.4.jar:0.23.4]
2020-08-04 23:41:09.609 [Broker-2-ZeebePartition-2] [Broker-2-zb-actors-2] ERROR io.zeebe.broker.system - Failed to install follower partition 2
java.lang.RuntimeException: Actor is closed
	at io.zeebe.util.sched.ActorJob.failFuture(ActorJob.java:166) ~[zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ClosedQueue.offer(ClosedQueue.java:84) ~[zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ClosedQueue.offer(ClosedQueue.java:15) ~[zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorTask.submit(ActorTask.java:95) ~[zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorControl.call(ActorControl.java:135) ~[zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.logstreams.impl.log.LogStreamImpl.getCommitPositionAsync(LogStreamImpl.java:105) ~[zeebe-logstreams-0.23.4.jar:0.23.4]
	at io.zeebe.engine.processor.AsyncSnapshotDirector.lambda$closeAsync$1(AsyncSnapshotDirector.java:108) ~[zeebe-workflow-engine-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:28) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:76) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.23.4.jar:0.23.4]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.23.4.jar:0.23.4]
2020-08-04 23:41:14.059 [Broker-2-TopologyManager] [Broker-2-zb-actors-1] DEBUG io.zeebe.broker.clustering - Received metadata change for 1, partitions {1=FOLLOWER, 2=LEADER, 3=FOLLOWER} terms {2=216}
```

</p>
</details>

This has resulted in our standalone gateway being in this state again (thought this was fixed in #4557 ):
```
Cluster size: 3
Partitions count: 3
Replication factor: 3
Gateway version: 0.23.4
Brokers:
  Broker 0 - prod-zeebe-zeebe-broker-0.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26501
    Version: 0.23.4
    Partition 1 : Leader
    Partition 2 : Follower
    Partition 3 : Leader
  Broker 1 - prod-zeebe-zeebe-broker-1.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26501
    Version: 0.23.4
    Partition 1 : Follower
    Partition 2 : Leader
    Partition 3 : Follower
  Broker 2 - prod-zeebe-zeebe-broker-2.prod-zeebe-zeebe-broker.prod.svc.cluster.local:26501
    Version: 0.23.4
    Partition 1 : Follower
    Partition 3 : Follower
```
I don't think these are related to this issue. But are these expected errors? If not, I can raise a separate issue for this.

**Comment by npepinpe** [comment]
Yes this is a separate issue - seems an unexpected exception (which is sort of expected - we tried to append asynchronously while transitioning from leader to follower, which is why you see the first error) causes us to be unable to install the follower service, which means you will not be getting snapshots and may eventually run out of disk space. I would restart that node as a workaround, but before please open a new issue with the logs.

**Comment by strawhat5** [comment]
Thanks Nicolas, I have created a separate issue.

**Comment by strawhat5** [comment]
Hey final update on this issue:

After reconfiguring one exporter and restarting all the brokers, we can see snapshot operations happening for all the partitions. Atomix log compaction is also happening for all the partitions now. Due to this the older `raft-partition*.log` files are now getting deleted, albeit gradually. I am expecting it would take a few days to delete all the older log-segments comprising of over 150GB of our data. 😅 

Thanks a lot @korthout @Zelldon and @npepinpe for root-causing the issue and fixing it! 🙇‍♂️  Please feel free to close this issue now or after the patch-release.

**Comment by korthout** [comment]
@Zelldon do you want to wait for the patch release or can we close this issue now?

**Comment by Zelldon** [comment]
Hey @strawhat5 ,

we released [0.23.5](https://github.com/zeebe-io/zeebe/releases/tag/0.23.5) [0.24.2](https://github.com/zeebe-io/zeebe/releases/tag/0.24.2), which includes the fix.

Check out the blog post:
https://zeebe.io/blog/2020/08/zeebe-0235-0242-release/

**Comment by strawhat5** [comment]
Thanks Chris! :)
